port to python3

kobbled opened the pull request:
With python finally dropping support for python2 next year it might be a good idea to patch for python3. Only potential issue is that newer version of python use the py launcher executable instead of calling python with 'python' or 'python3' so the .cmd file also needed to be updated, which will not work if python launcher is not installed.

kobbled committed:
port to python3

update to python 3

gavanderhoorn commented on ktransw.py:
```diff
@@ -243,7 +243,7 @@ def main():
                         logger.debug("Found {0} in '{1}'".format(hdr, hdr_dir))
 
-                    except ValueError as e:
+                    except (ValueError) as e:
```
A quick Google doesn't explain to me why the parentheses should be added here. Is this a Python3 best practice?

kobbled replied:
I believe it was throwing an error without it. I had to google that myself. Should double check though.

gavanderhoorn commented on ktransw.py:
```diff
@@ -312,7 +312,9 @@ def main():
             # preprocessed KAREL source file, not the original one.
-            sys.stdout.write(pstdout.replace(dname, os.path.dirname(kl_file)) + '\n')
+            pstdout = pstdout.decode('utf-8')
+            ktrans_out = pstdout.replace(dname, os.path.dirname(kl_file)) + '\n'
```
```suggestion
            ktrans_out = pstdout.decode('utf-8').replace(dname, os.path.dirname(kl_file)) + '\n'
```

Could be nice to chain this (and then remove the `pstdout = pstdout.decode('utf-8')` line above).

kobbled replied:
definitely. I think I was just testing and forgot to get rid of the extra line.

gavanderhoorn commented on ktransw.cmd:
```diff
@@ -1,17 +1,33 @@
+
 REM limitations under the License.
+
 REM
```
This seems to introduce a lot of empty lines. Is this a line-ending conflict?

kobbled replied:
Rossum.cmd was fine in my editor. ktransw.cmd had no carriage returns for me, just one line.

gavanderhoorn replied:
>  ktransw.cmd had no carriage returns for me, just one line

You may have a global setting for line-endings that is (trying to) overriding what the repository is configured for. See whether [Dealing with line endings](https://help.github.com/en/articles/dealing-with-line-endings) sheds some light on this.

gavanderhoorn reviewed (commented):
Thanks for the PR. A few in-line comments and a more general one.

Is the Python launcher really needed? I've not used it myself, but from what I gather from reading some sites it seems to be an optional part of Python 3 (it seems to not even always be included in an install). Calling `python.exe` should still work, it would depend on the `PATH` of the user which Python is then used.

I'm not against using the launcher, but would like to understand the situation a bit better.

kobbled commented:
Haven't tested using just 'python'. I still have a python2 and python3 on my machine so using just python will always give me python2. The py launcher does seem like a good idea to separate the two though. Don't know what the future entails for it. Need to investigate this further.

gavanderhoorn commented:
> Haven't tested using just 'python'. I still have a python2 and python3 on my machine so using just python will always give me python2.

So is `python2` in a directory that is before the `python3` directory in your `PATH`? That would/could explain this.

> The py launcher does seem like a good idea to separate the two though. Don't know what the future entails for it.

Python 2 doesn't have the launcher, so even if both intepreters are on the `PATH` it would only ever be found in the Python 3 location.

As I wrote above: I'm not against standardising on it, just want to understand it.

kobbled committed:
port to python3

Fix changes agreed to in PR#17

kobbled commented:
Thinking about it though I don't think the py launcher is standardizing. My python setup on windows is a bit of a mess for me so the py launcher helps out, but that is my fault for not using environments when needed. Embedding the python launcher into a batch file calling ktransw would be very confusing, and cause issues. I have reverted the changes to the cmd file.

kobbled commented:
I have updated the changes you suggested above to this branch!

kobbled committed:
remove blank lines from preprocessed temp file.

not removing blank lines caused from macros and includes cause the preprocessed output to be full of blank lines.

kobbled committed:
add ability to instantiate objects with `%class name('file.klc,'file.klh') preprocessor directive

- will create seperate .kl files for each object declared with %class.
-Each class requries a file defining the class '.klc' file and a header file '.klh', with an option for a type file '.klt'.
-See package 'bank_class' in github.com/kobbled/rossum_example_ws
-seperated gpp and ktrans into independent functions
-For classes a bunch of temp files are created (2 for each object), as two passes through gpp is required.  This may be hard on the system if there are a lot of object being created.
-see package ktransw-macros for information on how to create classes.

kobbled committed:
introduce hack for dealing with gpp short comings

-dynamic naming of preprocessor variables at compile time is not possible with the current configuration of gpp in ktransw.
-You cannot nest `string` and `quote` modes, or seperate individual lines of a function.
-docstring for 'remove_char' explains the dilema
-introduced a remove_char function to remove a `nostring` mode char from the file before running through preprocessor
-introduces extra pass through gpp. not good for speed, or memory usage.

gavanderhoorn commented:
Hi @kobbled. You seem to have pushed some additional commits to this branch.

That's OK, but I'm not sure they should be part of this PR.

I noticed you've been really making serious use of `gpp` and its capabilities. Could you enable *issues* on your fork of `ktransw_py`? I'd like to post a comment about your implementation of objects and classes.

kobbled commented:
@gavanderhoorn, I opened up the issues on the fork. Yes, it been a while since I've come back to this so I forgot this PR was still open and that it wasn't on a separate branch. If you want to close this, I can create a branch up to a reasonable commit and resubmit. With extra covid time I've been experimenting with gpp, and your libsimplemsg got me thinking it wouldn't be to difficult to implement crude objects into ktransw with some hand waving with gpp. I'm just getting around to understanding how to use a preprocessor in more depth than the typical uses you see in c.  There are definitely some not good practices I used to achieve this, but I just wanted to see if I could get it to work with the experience that I have.

kobbled committed:
enable recursive %class searching

-put gpp cycles in a recursive function in order to be able to process nested class objects.
-added one or two more gpp cycles. It is actually noticibly slower now.
-**see if there are any redundant gpp calls to spead up.
